lib/net/modules/dns.toit

```diff
@@ -1,29 +1,193 @@
-// Copyright (C) 2018 Toitware ApS. All rights reserved.
-// Use of this source code is governed by an MIT-style license that can be
-// found in the lib/LICENSE file.
-
-import monitor
+import binary show BIG_ENDIAN
+import net.modules.udp
 import net
 
-dns_lookup hostname:
-  id := null
-  while not id:
-    id = dns_lookup_ dns_resource_group_ hostname
-    // In LWIP's implementation of DNS there is a maximum number of
-    // outstanding DNS lookups at one time.  If we hit the limit, sleep a
-    // little and try later.
-    if not id: sleep --ms=10
-  state := monitor.ResourceState_ dns_resource_group_ id
-  state.wait
-  return net.IpAddress (dns_lookup_result_ dns_resource_group_ id)
+/**
+Look up a domain name.
+
+If given a numeric address like 127.0.0.1 it merely parses
+  the numbers without a network round trip.
+
+Does not currently cache results, so there is normally a
+  network round trip on every use.
+
+By default the server is "8.8.8.8" which is the Google DNS
+  service.
+
+Currently only works for IPv4, not IPv6.
+*/
+dns_lookup host/string --server/string="8.8.8.8" --timeout/Duration=DNS_DEFAULT_TIMEOUT -> net.IpAddress:
+  q := DnsQuery host
+  return q.get --server=server --timeout=timeout
+
+DNS_DEFAULT_TIMEOUT ::= Duration --s=20
+DNS_RETRY_TIMEOUT ::= Duration --s=1
+
+class DnsQuery:
+  static A     ::= 1  // A address.
+  static CNAME ::= 5  // Canonical name.
+  static CLASS ::= 1  // The Internet class.
+  id/int
+  name/string
+  hosts/Map ::= {"localhost": "127.0.0.1"}
+
+  constructor .name:
+    id = random 0x10000
+
+  /**
+  Look up a domain name.
+
+  If given a numeric address like "127.0.0.1" it merely parses
+    the numbers without a network round trip.
+
+  Does not currently cache results, so there is normally a
+    network round trip on every use.
+
+  By default the server is "8.8.8.8" which is the Google DNS
+    service.
+
+  Currently only works for IPv4, not IPv6.
+  */
+  get --server/string="8.8.8.8" --timeout/Duration=DNS_DEFAULT_TIMEOUT -> net.IpAddress:
+    if ip_string_:
+      return net.IpAddress.parse name
+    if hosts.contains name:
+      return net.IpAddress.parse hosts[name]
+
+    query := create_query_
+    socket := udp.Socket
+    with_timeout timeout:
+      try:
+        socket.connect
+          net.SocketAddress
+            net.IpAddress.parse server
+            53                             // DNS UDP port.
+
+        retry_timeout := DNS_RETRY_TIMEOUT
+
+        // Resend the query with exponential backoff until the outer timeout
+        // expires.
+        while true:
+          socket.write query
+
+          answer := null
+          catch:
```

Comment:
Here you swallow all stack traces. Maybe think about how to make that more selective (timeout).